### PR TITLE
allow to return a zero value patch pointer

### PR DESCRIPTION
**Pull request by xh3b4sd**
IMO it should be possible to just return `nil` in the [`New*Patch` methods](https://github.com/giantswarm/node-operator/blob/cebb26de5db6b13cfaffa662c18757cd58e6a505/service/nodeconfig/v1/resource/node/delete.go#L13-L16). This PR makes this possible. Hashtag boyscouting.

### CODE REVIEW

File: framework/framework.go [hunk 1]
```diff
@@ -371,10 +371,6 @@ func ProcessDelete(ctx context.Context, obj interface{}, resources []Resource) e
 			if err != nil {
 				return microerror.Mask(err)
 			}
-
-			if patch == nil {
-				return microerror.Maskf(executionFailedError, "patch must not be nil")
-			}
```
Comment (xh3b4sd) [inline framework/framework.go]: I changed this in `ProcessDelete` and `ProcessUpdate`. `ProcessCreate` is going to be replaced in another PR.

File: framework/patch.go [hunk 2]
```diff
@@ -24,6 +24,18 @@ func NewPatch() *Patch {
...
+	p.data[patchCreate] = create
+}
+
+func (p *Patch) SetDeleteChange(delete interface{}) {
+	p.data[patchDelete] = delete
+}
+
+func (p *Patch) SetUpdateChange(update interface{}) {
+	p.data[patchUpdate] = update
+}
```
Comment (xh3b4sd) [inline framework/patch.go]: Reformatting this to be more consistent. Also public methods first.

File: framework/patch_dispatch_change_test.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"testing"
+)
+
+func Test_Framework_Resource_PatchDispatchChange(t *testing.T) {
```
Comment (xh3b4sd) [inline framework/patch_dispatch_change_test.go]: This should make sure dispatching the patch changes work as expected. We did not seem to have such tests so far.

File: framework/patch_no_panic_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
+package framework
+
+import (
+	"context"
+	"testing"
+)
+
+func Test_Framework_Resource_PatchNoPanic(t *testing.T) {
```
Comment (xh3b4sd) [inline framework/patch_no_panic_test.go]: These tests make sure nil patches do not panic.

**Comment by xh3b4sd** [comment]
> I'd rather make the Patch struct a value.

What does that mean? Do you mean no pointer? I do not think the setting of arbitrary information will then work out. I think it is ok as it is. The only thing I want to be able is to return `nil`. Because this is the language specific default behaviour in this case. What do you have against my proposal?

**Comment by kopiczko** [comment]
> I do not think the setting of arbitrary information will then work out

AFAIK it will. map holds a pointer, so you can safely copy it. See https://play.golang.org/p/b_0HN1KewyP

**Comment by kopiczko** [comment]
> What do you have against my proposal?

Nothing agains. I think we can do a little better. That's all.

**Comment by xh3b4sd** [comment]
I would go with this since there are some improvements and additional tests. IMO it is the right thing to have the pointers. Being able to just return nil is very nice and there are no real drawbacks in the current implementation. When you think patch should rather be a value you can do this still. I would disagree though, but not veto.
